# Update Plethora to CC:Tweaked version 1.115.1 minecraft 1.20.1

Greeley opened the pull request:
Usually assemble my own modpacks, I hate not having Plethora around but am using CC:Tweaked 1.115.1 for compatibility between other mods. These changes update Plethora to 1.115.1

## KeyboardModuleItem
- updated `useOnComputer` to use the new CC:Tweaked paradigm for opening a menu.
- changed `isComputerUsableWithRange` check due to `CommandComputerBlockEntity` being removed

## KeyboardScreenHandlerFactory
- minor change to `CreateMenu` to match new signature of `ComputerMenuWithoutInventory`

## MediaItemMeta
- This one was hard with `MediaProviders` being removed. Attempt to cast to `IMedia`, if not cast to new `RecordMedia` Instance.
- note: I'm not entirely sure how to test this or what hits this code. I'd be happy to test if anyone has a scenario for me.

## NeuralComputer
- Change to match `SeverComputer` signature change
- level and position are private, calls set position instead.

## NeuralPocketAccess
- `getLevel` is now required to implement this interface
- `getPosition` is now required to implement this interface
- `getUpgrades` has been deprecated I noticed it returned empty anyway
- implemented `setUpgrade` and `getUpgrade` but I'm not sure they do anything/should be. Feedback on this is appreciated.
- changed upgrades return to `Map.of()` so we don't need `Collections` import

## PocketUpgradeModule
- changed upgrade check as `getUpgrades` is deprecated.

Greeley committed:
# Update Plethora to CC:Tweaked version 1.115.1 minecraft 1.20.1

Usually assemble my own modpacks, I hate not having Plethora around but am using CC:Tweaked 1.115.1 for compatibility between other mods. These changes update Plethora to 1.115.1

## KeyboardModuleItem
- updated `useOnComputer` to use the new CC:Tweaked paradigm for opening a menu.
- changed `isComputerUsableWithRange` check due to `CommandComputerBlockEntity` being removed

## KeyboardScreenHandlerFactory
- minor change to `CreateMenu` to match new signature of `ComputerMenuWithoutInventory`

## MediaItemMeta
- This one was hard with `MediaProviders` being removed. Attempt to cast to `IMedia`, if not cast to new `RecordMedia` Instance.
- note: I'm not entirely sure how to test this or what hits this code. I'd be happy to test if anyone has a scenario for me.

## NeuralComputer
- Change to match `SeverComputer` signature change
- level and position are private, calls set position instead.

## NeuralPocketAccess
- `getLevel` is now required to implement this interface
- `getPosition` is now required to implement this interface
- `getUpgrades` has been deprecated I noticed it returned empty anyway
- implemented `setUpgrade` and `getUpgrade` but I'm not sure they do anything/should be. Feedback on this is appreciated.
- changed upgrades return to `Map.of()` so we don't need `Collections` import

## PocketUpgradeModule
- changed upgrade check as `getUpgrades` is deprecated.

SquidDev commented on build.gradle.kts:
```diff
@@ -69,7 +69,7 @@ repositories {
+  maven("https://maven.squiddev.cc/") {
     content {
       includeGroup("cc.tweaked")
       includeModule("org.squiddev", "Cobalt")
```
```suggestion
```

Think this can go now!

SquidDev commented on src/main/java/io/sc3/plethora/integration/computercraft/meta/item/MediaItemMeta.kt:
```diff
@@ -1,12 +1,16 @@
   override fun getMeta(target: ItemStack): Map<String, *> {
-    val media = MediaProviders.get(target) ?: return emptyMap<String, Any>()
+
+    val media = if(target.item is IMedia) target.item as IMedia else RecordMedia.INSTANCE
```
Think you can use `dan200.computercraft.api.media.MediaLookup` here instead:

```suggestion
    val media = MediaLookup.get().find(target, null)
```

Greeley replied:
With this change There's a possibility that the `media` variable could be null.

Would it be preferred to wrap in a null check and then return an empty map if there's no metadata, or use nullable field access to provide the current map but each item set to null?

Lemmmy replied:
You can probably just do this:

```
val media = MediaLookup.get().find(target, null) ?: return emptyMap()
```

Greeley replied:
This is good, just had to give it the types.
Thanks :)

SquidDev commented on src/main/java/io/sc3/plethora/gameplay/modules/keyboard/KeyboardModuleItem.kt:
```diff
@@ -118,7 +118,10 @@ class KeyboardModuleItem(settings: Settings) : ModuleItem("keyboard", settings)
+
+
+
+      if (blockEntity.family == ComputerFamily.COMMAND && !blockEntity.serverComputer?.level?.server?.areCommandBlocksEnabled()!!) {
```
I think we can use `checkUsable` here?

```suggestion
      if (!blockEntity.family.checkUsable(player))`
```

Greeley replied:
Yeah that seems simpler to understand as well, I'll make the change and test it out before committing.
Thanks for the review.

Greeley replied:
I've updated this and experienced no problems with they keyboard in my world.

Greeley committed:
# Update Plethora to CC:Tweaked version 1.115.1 minecraft 1.20.1

## KeyboardModuleItem
- changed `isComputerUsableWithRange` check to check if blockEntity is not usable by player.

Greeley committed:
Update src/main/java/io/sc3/plethora/integration/computercraft/meta/item/MediaItemMeta.kt

Yeah! this one confused me for a second, but this looks way better. I was worried about what would happen if the else branch was executed.

Co-authored-by: Jonathan Coates <git@squiddev.cc>

Greeley committed:
Update build.gradle.kts

Thanks!

Co-authored-by: Jonathan Coates <git@squiddev.cc>

Greeley committed:
# Update Plethora to CC:Tweaked version 1.115.1 minecraft 1.20.1

## MediaItemMeta
- updated `getMeta` uses `MediaLookup` to get metadata of a media item. If item is not a media item an emptyMap<String, String> is given as the metadata instead.

SquidDev committed:
Remove unused imports

SquidDev committed:
Correctly set label

Lemmmy commented:
Thanks for working on this @Greeley and for reviewing @SquidDev!

SquidDev committed:
Use famiyl.checkUsable

Greeley commented:
> Thanks for working on this @Greeley and for reviewing @SquidDev!

Thanks @Lemmmy I had fun, it's a nice codebase and am happy to contribute back to the community. 

@SquidDev Thanks for reviewing! I Appreciate the thoroughness.
